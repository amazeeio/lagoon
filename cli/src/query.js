// @flow

import path from 'path';
import os from 'os';
import url from 'url';
import R from 'ramda';
import { fileExists, readFile } from './util/fs';
import request from './util/request';

type QLQueryArgs = {
  endpoint?: string,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function runGQLQuery(args: QLQueryArgs): Promise<Object> {
  const {
    endpoint = process.env.API_URL || 'https://api.amazee.io/graphql',
    query,
    variables,
    headers: customHeaders = {},
    pretty = false,
  } = args;

  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = path.join(os.homedir(), '.ioauth');
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const token = await readFile(tokenFile);
      headers.Authorization = `Bearer ${encodeURIComponent(token.toString())}`;
    }
  }

  const { hostname, path: pathname, port: urlPort, protocol = 'https:' } = url.parse(endpoint);

  if (hostname == null) {
    throw new Error('Hostname required');
  }

  const body = JSON.stringify(
    {
      query,
      variables,
    },
    null,
    pretty ? 2 : 0,
  );

  const protocolPorts = {
    'https:': 443,
    'http:': 80,
  };

  const options = {
    hostname,
    path: pathname,
    port: urlPort === null ? R.propOr(443, protocol)(protocolPorts) : Number(urlPort),
    method: 'POST',
    headers,
    body,
    rejectUnauthorized: false,
  };

  return request(options);
}

export default runGQLQuery;
