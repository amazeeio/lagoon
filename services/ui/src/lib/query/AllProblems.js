import gql from 'graphql-tag';

export default gql`
  {
    allGroups {
      name
      projects {
        name
        id
        environments(type: PRODUCTION) {
          name
          id
          problems {
            id
            identifier
            severity
            source
          }
        }
      }
    }
  }
`;
