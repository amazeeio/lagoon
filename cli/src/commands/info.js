// @flow

import { table } from 'table';
import { red } from 'chalk';
import R from 'ramda';

import gql from '../gql';
import { runGQLQuery } from '../query';
import {
  printErrors,
  printNoConfigError,
  printGraphQLErrors,
} from '../printErrors';

import typeof Yargs from 'yargs';
import type { BaseArgs } from './index';

const tableConfig = {
  columns: {
    // $FlowIssue: Flow doesn't understand numbers as keys https://github.com/facebook/flow/issues/380
    0: {
      alignment: 'left',
      minWidth: 15,
    },
    // $FlowIssue: Flow doesn't understand numbers as keys https://github.com/facebook/flow/issues/380
    1: {
      alignment: 'left',
      minWidth: 15,
    },
  },
};

// Common filter
const onlyValues = ([, value]: [string, string]) =>
  value != null && value !== '';

const name = 'info';
const description = 'Show info about sites or sitegroups';

export async function setup(yargs: Yargs): Promise<Object> {
  return yargs
    .usage(`$0 ${name} [site@branch] - ${description}`)
    .options({
      sitegroup: {
        demandOption: false,
        describe: 'Override the currently configured sitegroup (.amazeeio.yml)',
        type: 'string',
      },
    })
    .alias('s', 'sitegroup')
    .example(
      `$0 ${name}`,
      'Show information about sitegroup in the .amazeeio.yml config file',
    )
    .example(
      `$0 ${name} mysite`,
      'Show information about site "mysite" (only works if the site only has a single branch) (sitegroup as stated by config)',
    )
    .example(
      `$0 ${name} mysite@prod`,
      'Show information about branch "prod" of site "mysite" (sitegroup as stated by config)',
    )
    .example(
      `$0 ${name} -s mysitegroup mysite`,
      'Show information about site "mysite" in sitegroup "somesitegroup"',
    ).argv;
}

type SiteGroupInfoArgs = {
  sitegroup: string,
  clog: typeof console.log,
  cerr: typeof console.error,
};

export async function sitegroupInfo({
  sitegroup,
  clog,
  cerr,
  }: SiteGroupInfoArgs): Promise<number> {
  const query = gql`
    query querySites($sitegroup: String!) {
      siteGroupByName(name: $sitegroup) {
        gitUrl
        siteGroupName
        slack {
          webhook
          channel
          informStart
          informChannel
        }
        client {
          clientName
        }
        sites {
          siteName
          siteBranch
        }
      }
    }
  `;

  const result = await runGQLQuery({
    query,
    variables: { sitegroup },
  });

  const { errors } = result;
  if (errors != null) {
    return printGraphQLErrors(cerr, ...errors);
  }

  const sitegroupData = R.path(['data', 'siteGroupByName'])(result);

  if (sitegroupData == null) {
    return printErrors(clog, `No sitegroup '${sitegroup}' found`);
  }

  const sites = R.compose(
    R.map(({ siteName, siteBranch }) => `${siteName}:${siteBranch}`),
    R.pathOr([], ['data', 'siteGroupByName', 'sites']),
  )(result);

  const formatSlack = (slack) => {
    if (slack == null) {
      return '';
    }

    const webhook = R.prop('webhook', slack);
    const channel = R.prop('channel', slack);

    return `${channel} -> ${webhook}`;
  };

  const tableBody = [
    ['SiteGroup Name', R.prop('siteGroupName', sitegroupData)],
    ['Git Url', R.prop('gitUrl', sitegroupData)],
    ['Slack', formatSlack(R.prop('slack', sitegroupData))],
    ['Sites', R.join(', ', sites)],
  ];

  const tableData = R.filter(onlyValues)(tableBody);

  clog(`I found following information for sitegroup '${sitegroup}':`);
  clog(table(tableData, tableConfig));

  return 0;
}

type SiteInfoArgs = {
  site: string,
  branch?: string,
  sitegroup: string,
  clog: typeof console.log,
  cerr: typeof console.error,
};

export async function siteInfo({
  sitegroup,
  site,
  branch,
  clog,
  cerr,
  }: SiteInfoArgs): Promise<number> {
  // site[@branch]
  const siteBranchStr = `${site}${branch != null ? `@${branch}` : ''}`;

  const query = gql`
    query querySites($sitegroup: String!) {
      siteGroupByName(name: $sitegroup) {
        gitUrl
        sites {
          id
          siteHost
          siteName
          siteBranch
          siteEnvironment
          serverInfrastructure
          serverIdentifier
          serverNames
          webRoot
          drupalVersion
          SSLCertificateType
          FPMProfile
          domains
          redirectDomains
          redirects
          uid
          dbUser
          cron {
            type
            minute
          }
          customCron
          envVariables
          noPrefixenvVariables
          phpValues
          phpAdminFlags
          xdebug
          nginxSitespecific
          nginxSiteconfig
          solrEnabled
          redisEnabled
          sshKeys
          phpVersion
          redirectToHttps
          ensure
          upstreamURL
          apc
          basicAuth {
            username
          }
        }
      }
    }
  `;

  const result = await runGQLQuery({
    query,
    variables: { sitegroup },
  });

  const { errors } = result;
  if (errors != null) {
    return printGraphQLErrors(cerr, ...errors);
  }

  // There might be a case where I just have one site + branch,
  // so we will just filter for branch, if a branch was actually
  // given. If there are multiple matches, we will tell the user
  // anyways
  const matchesSiteAndBranch = ({ siteName, siteBranch }) => {
    if (branch != null && siteBranch !== branch) {
      return false;
    }
    return siteName === site;
  };

  const nodes = R.compose(
    R.filter(matchesSiteAndBranch),
    R.pathOr([], ['data', 'siteGroupByName', 'sites']),
  )(result);

  if (nodes.length === 0) {
    clog(red(`No site '${site}' found in sitegroup '${sitegroup}'`));
    return 0;
  }

  // For the case if there was no branch name given to begin with
  if (nodes.length > 1) {
    clog(
      'I found multiple sites with the same name, but different branches, maybe try following parameter...',
    );
    R.forEach(({ siteName, siteBranch }) =>
      clog(`-> ${siteName}@${siteBranch}`),
    )(nodes);
    return 0;
  }

  clog(`I found following information for '${sitegroup} -> ${siteBranchStr}':`);

  // nodes only contains one element, extract it
  const [node] = nodes;

  const formatCron = (cron) => {
    if (cron == null) {
      return '';
    }

    return `Type: ${cron.type} - Minute: ${cron.minute}`;
  };

  const formatArray = (arr) => {
    if (arr == null) {
      return '';
    }

    return R.join(', ', arr);
  };

  // We want to list these fields from the node result
  const tableBody = [
    ['ID', R.prop('id', node)],
    ['Site Name', R.prop('siteName', node)],
    ['Site Branch', R.prop('siteBranch', node)],
    ['Site Env', R.prop('siteEnvironment', node)],
    ['User Id', R.prop('uid', node)],
    ['Server Names', R.prop('serverNames', node)],
    ['Webroot', R.prop('webRoot', node)],
    ['Domains', formatArray(R.prop('domains', node))],
    ['Redirect Domains', formatArray(R.prop('redirectDomains', node))],
    ['SSL Certificate Type', R.prop('SSLCertificateType', node)],
    ['Cron', formatCron(R.prop('cron', node))],
    ['Solr Enabled', R.prop('solrEnabled', node)],
  ];

  const tableData = R.filter(onlyValues)(tableBody);

  clog(table(tableData, tableConfig));

  return 0;
}

type Args = BaseArgs & {
  sitegroup: ?string,
};

export async function run(args: Args): Promise<number> {
  const { config, clog, cerr } = args;

  // FIXME: doesn't handle empty config file case correctly
  if (config == null) {
    return printNoConfigError(cerr);
  }

  const [siteAndBranch] = args._.slice(1);
  const sitegroup = args.sitegroup || config.sitegroup;

  if (siteAndBranch == null) {
    return sitegroupInfo({ sitegroup, clog, cerr });
  }

  const [site, branch] = siteAndBranch.split('@');

  return siteInfo({
    site,
    branch,
    sitegroup,
    clog,
    cerr,
  });
}

export default {
  setup,
  name,
  description,
  run,
};
