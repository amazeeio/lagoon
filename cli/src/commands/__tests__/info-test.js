// @flow

import { runGQLQuery } from '../../query';
import { siteInfo, sitegroupInfo } from '../info';

jest.mock('../../query');

const _mock = (mockFn: any): JestMockFn => mockFn;

const mockErrorResponse = {
  errors: [{ message: 'something something error' }],
};

describe('siteGroupInfo', () => {
  const mockResponse1 = {
    data: {
      siteGroupByName: {
        gitUrl: 'sitegroup.git',
        siteGroupName: 'mysitegroup',
        slack: {
          webhook: 'https://slack-webhook.something',
          channel: 'myslack',
        },
        client: {
          clientName: 'me',
        },
        sites: [
          {
            siteName: 'site1',
            siteBranch: 'dev',
          },
          {
            siteName: 'site1',
            siteBranch: 'prod',
          },
        ],
      },
    },
  };

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockErrorResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await sitegroupInfo({
      sitegroup: 'some_sitegroup',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show error on missing sitegroup', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() => Promise.resolve({}));

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await sitegroupInfo({
      sitegroup: 'not_existing',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should list found information for given sitegroup', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockResponse1),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await sitegroupInfo({
      sitegroup: 'mysitegroup',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});

describe('siteInfo', () => {
  // Case for testing ambiguous sitenames
  const mockResponse1 = {
    data: {
      siteGroupByName: {
        sites: [
          {
            siteName: 'site1',
            siteBranch: 'dev',
          },
          {
            siteName: 'site1',
            siteBranch: 'prod',
          },
        ],
      },
    },
  };

  // Testdata for proper response without ambiguity
  const mockResponse2 = {
    data: {
      siteGroupByName: {
        sites: [
          {
            siteName: 'site1',
            siteBranch: 'dev',
            siteEnvironment: 'development',
            uid: 'uid',
            serverNames: ['servername1'],
            webRoot: 'webroot',
            domains: ['domain1', 'domain2'],
            redirectDomains: ['redomain1', 'redomain2'],
            SSLCertificateType: 'sslcerttype',
            cron: { type: 'cron', minute: 0 },
            solrEnabled: false,
          },
          {
            siteName: 'site2',
            siteBranch: 'prod',
          },
        ],
      },
    },
  };

  it('should detect ambiguity and propose more specific parameters', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockResponse1),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await siteInfo({
      sitegroup: 'some_sitegroup',
      site: 'site1',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should display error, if GraphQL sends error messages', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockErrorResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await siteInfo({
      sitegroup: 'some_sitegroup',
      site: 'not_existing',
      clog,
      cerr,
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show table with information about sitegroup', async () => {
    _mock(runGQLQuery).mockImplementationOnce(() =>
      Promise.resolve(mockResponse2),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await siteInfo({
      sitegroup: 'some_sitegroup',
      site: 'site1',
      branch: 'dev',
      clog,
      cerr,
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
