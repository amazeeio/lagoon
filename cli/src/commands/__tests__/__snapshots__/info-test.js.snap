// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`siteGroupInfo should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The amazee.io API returned an error:[39m",
  ],
  Array [
    "[31m{
  \\"message\\": \\"something something error\\"
}[39m",
  ],
]
`;

exports[`siteGroupInfo should list found information for given sitegroup 1`] = `
Array [
  Array [
    "I found following information for sitegroup 'mysitegroup':",
  ],
  Array [
    "╔════════════════╤════════════════════════════════════════════╗
║ SiteGroup Name │ mysitegroup                                ║
╟────────────────┼────────────────────────────────────────────╢
║ Git Url        │ sitegroup.git                              ║
╟────────────────┼────────────────────────────────────────────╢
║ Slack          │ myslack -> https://slack-webhook.something ║
╟────────────────┼────────────────────────────────────────────╢
║ Sites          │ site1:dev, site1:prod                      ║
╚════════════════╧════════════════════════════════════════════╝
",
  ],
]
`;

exports[`siteGroupInfo should show error on missing sitegroup 1`] = `
Array [
  Array [
    "[31mNo sitegroup 'not_existing' found[39m",
  ],
]
`;

exports[`siteInfo should detect ambiguity and propose more specific parameters 1`] = `
Array [
  Array [
    "I found multiple sites with the same name, but different branches, maybe try following parameter...",
  ],
  Array [
    "-> site1@dev",
  ],
  Array [
    "-> site1@prod",
  ],
]
`;

exports[`siteInfo should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The amazee.io API returned an error:[39m",
  ],
  Array [
    "[31m{
  \\"message\\": \\"something something error\\"
}[39m",
  ],
]
`;

exports[`siteInfo should show table with information about sitegroup 1`] = `
Array [
  Array [
    "I found following information for 'some_sitegroup -> site1@dev':",
  ],
  Array [
    "╔══════════════════════╤════════════════════════╗
║ Site Name            │ site1                  ║
╟──────────────────────┼────────────────────────╢
║ Site Branch          │ dev                    ║
╟──────────────────────┼────────────────────────╢
║ Site Env             │ development            ║
╟──────────────────────┼────────────────────────╢
║ User Id              │ uid                    ║
╟──────────────────────┼────────────────────────╢
║ Server Names         │ servername1            ║
╟──────────────────────┼────────────────────────╢
║ Webroot              │ webroot                ║
╟──────────────────────┼────────────────────────╢
║ Domains              │ domain1, domain2       ║
╟──────────────────────┼────────────────────────╢
║ Redirect Domains     │ redomain1, redomain2   ║
╟──────────────────────┼────────────────────────╢
║ SSL Certificate Type │ sslcerttype            ║
╟──────────────────────┼────────────────────────╢
║ Cron                 │ Type: cron - Minute: 0 ║
╟──────────────────────┼────────────────────────╢
║ Solr Enabled         │ false                  ║
╚══════════════════════╧════════════════════════╝
",
  ],
]
`;
